SIL.Archiving/IMDI/IMDIArchivingDlgViewModel.cs

```diff
@@ -164,200 +125,93 @@ protected override string FileGroupDisplayMessage(string groupKey)
 		}
 
-		/// <summary></summary>
-		/// <returns></returns>
+
+		/// ------------------------------------------------------------------------------------
 		public override string GetMetadata()
 		{
 			return _imdiData.BaseImdiFile.ToString();
 		}
 
 		/// ------------------------------------------------------------------------------------
 		/// <summary>Launch Arbil or Lamus or whatever</summary>
-		/// <remarks>need custom launcher here because Arbil is a java program, with no executable on linux</remarks>
+		/// <remarks>Need custom launcher here because Arbil is a java program, with no
+		/// executable on linux</remarks>
 		/// ------------------------------------------------------------------------------------
-		internal override void LaunchArchivingProgram()
+		public override void LaunchArchivingProgram()
 		{
-			if (string.IsNullOrEmpty(PathToProgramToLaunch) || !File.Exists(PathToProgramToLaunch))
+			if (IsNullOrEmpty(PathToProgramToLaunch) || !File.Exists(PathToProgramToLaunch))
 				return;
 
 			// if it is a .jar file, open with java
-			var exePath = (PathToProgramToLaunch.EndsWith(".jar")) ? "java" : PathToProgramToLaunch;
-			var args = string.Empty;
+			var exePath = PathToProgramToLaunch.EndsWith(".jar") ? "java" : PathToProgramToLaunch;
+			var args = Empty;
 			if (exePath == "java")
 			{
 				// are there additional command line parameters for this program?
-				if (PathToProgramToLaunch.ToLower().Contains("arbil"))
-					args = string.Format(ArchivingPrograms.ArbilCommandLineArgs, PathToProgramToLaunch);
-				else
-					args = PathToProgramToLaunch;
+				args = PathToProgramToLaunch.ToLower().Contains("arbil") ?
+					Format(ArchivingPrograms.ArbilCommandLineArgs, PathToProgramToLaunch) :
+					PathToProgramToLaunch;
 			}
 
-			try
-			{
-				var prs = new Process { StartInfo = { FileName = exePath, Arguments = args } };
-				prs.Start();
-			}
-			catch (Exception e)
-			{
-				ReportError(e, string.Format(LocalizationManager.GetString("DialogBoxes.ArchivingDlg.StartingIMDIErrorMsg",
-					"There was an error attempting to open the archive package in {0}."), PathToProgramToLaunch));
-			}
+			var prs = new Process { StartInfo = { FileName = exePath, Arguments = args } };
+
+			LaunchArchivingProgram(prs);
 		}
 
-#region Create IMDI package in worker thread
+#region Create IMDI package asynchronously
 
-		/// <summary></summary>
-		public override bool CreatePackage()
+		/// ------------------------------------------------------------------------------------
+		public override async Task<string> CreatePackage(CancellationToken cancellationToken)
 		{
-			IsBusy = true;
-
 			// check for missing data that is required by Arbil
 			var success = _imdiData.SetMissingInformation();
 
 			// write the xml files
 			if (success)
-				success = _imdiData.CreateIMDIPackage();
+				success = _imdiData.CreateIMDIPackage(); // REVIEW: Should this also be awaited?
```

Comment:
yeah good catch, this is where this gets hard, because best practice would be to go all the way down and make everything async, but there's a point where that's not going to work, so I'd just wrap this in a Task run and call it. That said the task will keep running even if cancelled, so that's a lot harder to work around, passing the cancelation token into Task.Run really only prevents starting the task if it's already cancelled. We would have to rewrite `CreateIMDIPackage` to accept the token and do something with it.

Reply:
In reality, most of this stuff is pretty fast on modern hardware, so unless you have a lot of files or some really big files, it's probably going to finish before you've had much of a chance to think about canceling.